Deploy release

arian487 opened the pull request:
Define a deploy and release strategy.  This fixes https://github.com/Recidiviz/recidiviz-data/issues/71

colincadams commented on .travis.yml:
```diff
@@ -1,27 +1,60 @@
+- openssl aes-256-cbc -K $encrypted_f7b0b16e1068_key -iv $encrypted_f7b0b16e1068_iv
+  -in secrets.tar.enc -out secrets.tar -d
+- tar xvf secrets.tar
+- export CLOUD_SDK_REPO="cloud-sdk-$(lsb_release -c -s)" && echo "deb http://packages.cloud.google.com/apt
```
nit: I know the lines were long, but I found this section much more readable before with each command on a single line.

If we want to maintain a max line length, then I think we should still start each following command on it's own line. We could just make each one it's own yaml entry if that is preferred

arian487 replied:
Oh I didn't even do that, looks like the travis script automatically did that.

arian487 replied:
I've fixed it now.

arian487 commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
+and verify that it is happy by looking at the monitoring page [TODO #59](https://github.com/Recidiviz/recidiviz-data/issues/59) and also checking
+the logs for errors.
+4. For the next two days periodically check to make sure the build is happy and monitoring is good for all regions
+including the test region.  Pay special attention to the regions that are going to be promoted to prod
```
We may have some regions that are fake/constant in staging for testing purposes and some regions that have recently had scrapers written for them that go through the short trial period before being moved from the staging cron to the main cron.

arian487 commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+1. Verify that the tests are all passing in [Travis](https://travis-ci.org/Recidiviz/recidiviz-data)
+2. Merge branch `master` into `staging`.  This will deploy to our staging environment and accordingly promote the scrapers to cron that are
+set to be deployed later in the week.
```
Its just markdown, there is no line break in the actual doc.

jessex replied:
That's weird. The source for the file definitely looks like we accidentally hit enter between "are" and "set". It renders fine, but it doesn't need to be sent down to line 132.

arian487 replied:
I see it now, fixed.

arian487 commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
-`gcloud app deploy`. This will upload the full project to the cloud and enable it as the current version in your environment.
+### Deployment
+
+### Depoying a scraper
```
Fixed.

arian487 commented on .travis.yml:
```diff
@@ -1,27 +1,60 @@
+  on: staging
 notifications:
   slack:
-    secure: RRx5Klgwfr2ggUxugs2nBOt2atYoKDrv+RfVY2OeINsZqPeZoxpfKntpIctfdUfPIZofyNUfPFfIFjZdNWBwXHKTu1/3KkCslBDUoPW9En1np5RTrVUmZJJy7no+J19PuDUDQYcDRCcEw41PbjcE7NPfK12bTI7jCGKw3ZTX1tLWlXDz70BOIsJxW3gLrgopFo0mNGWKbIVBU4VeUxhOpxB3RzegtCNrTIcfiEpnuNcyNvzhTpzlQdinW4rXQjv5p7oxCEkMGTfywtVrUVLfrSMCCVde1rv2RUBJPmShmFbJjfWuozMMlVK82vXMjj+z3HrIYPAYMmbVM9zqn/SDjdwzfzbmldol8yKcW4DTmlH4phfw9qWGa/UdpcHPT5vUoNz ...
```
Thats weird, the travis sceript took that out...I'll add it back in.

arian487 replied:
Fixed.

jessex commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
+### Deployment
+
+### Deploying a scraper
+To deploy a scraper, simply edit the [staging cron](staging/cron.yaml).  The release engineer
```
It's not 100% explicit, but it seems implied that this process involves checking in a change to `staging/cron.yaml` to add a new scraper, and then checking in a change to remove that scraper from `staging/cron.yaml` and then add it to `cron.yaml`. Is that right? If so, it feels like a good amount of churn.

I wonder if there's a way to change the branch structure to avoid so much churn directly on master. We check the first change into master, merge it into staging, then check another change directly into staging, then merge that into production. This means that production and staging do not look exactly like master but behind by some number of commits. It's also not clear how the changes find their way from production back to master without yet another merge, which seems like a potential danger. They both include commits -- to move scrapers from one `cron.yaml` to another -- that were _never_ written to master. I think if we are going to move towards dedicated release branches, we should not have engineers manually pushing commits to those release branches that never go through master. I think that makes the release workflow a bit hairier than it should be, reduces the reliability of our CI pipeline on master by ensuring it is not _exactly_ what's in production (even if only a small amount), and increases the number of potential bugs by increasing inconsistency.

I think a better approach might be to consolidate down to a single cron.yaml file but find the ability to suppress all but a whitelisted set of entries in that cron on stage. If that hypothetical whitelist is maintained in code then we have to figure out how to not have exactly the same set of concerns, but maybe that whitelist could be managed through some direct API call to GCP performed on deploy through Travis? One way would be to pause/disable the _queues_ that are available through the Cloud Task Queue API and let the crons all run but immediately no-op. That would be a bit noisy in the logs, admittedly. That would be trading off operational complexity for a cleaner commit log and branch structure. There could be better approaches with a more obviously positive trade.

arian487 replied:
Hm I definitely see the concerns you're bringing up.  I'm not in favour personally of using the Queue method because I think it can make the logs a bit confusing.  Someone might be trying to debug a region only to see that it didn't run, and then be confused why it didn't run.  I like the idea of letting the crons handle it...let me think on this and get back to you tomorrow.

arian487 replied:
Ok I may have just mentally convinced myself not to flow the mid week release out from master.  That would actually be quite dangerous as it would introduce unverified code into prod.  I think we would have to put into staging directly if we wanted to respect the flow.  I think that wouldn't be so horrible even if we didn't merge back from master as long as ensure the only changes that happen directly on the release branches is the cron changes.

jessex replied:
Almost every release strategy I have experience with in Git goes roughly like this: 1) cut a new tag/release when tested and ready, 2) deploy that release to staging and test and monitor, then 3) deploy that release to production. It's how Git is designed to work, if I can be so bold (obviously Git can work in a variety of ways, but this is the natural path). 

It seems like the two reasons that's not the current suggestion, as far as I can tell, are because we're not sure Travis can deploy from tags and we are trying to ensure that staging only runs a very limited but dynamic set of scrapers on schedule. 

The first one should be solveable with elbow grease. The second one should be mitigatable via tooling. For one, releases that don't involve new scrapers should have none of the complexity discussed here and should have a simpler strategy. For two, I still think there should be a way to get the testing we want without jumping through what seems to me like a counterintuitive release flow. Do we need to automatically schedule each new scraper, or are manual runs okay? If automatic is required, it's only for a couple of days, so do we need to use hard coded yaml for staging or is there a different mechanism? For example, could we put together an API call that is triggered in stage that specifically finds new scrapers since the last release and runs them? That specific idea may not be great, but I think the bigger point stands.

arian487 replied:
Ok I just mentally convinced myself that is a bad idea since it introduces unverified commits to staging and then prod.  One solution is to remove the cron deploy process from the general release process and have the release engineer manually deploy that.  It would be the same process of updating master and pushing those to the respective branches, except the engineer will have to run a deploy script in master on mondays and wednesdays which deploys the correct cron job.

That way any actual code is still running on the correct versions.  Thoughts?

arian487 replied:
Automatic runs are definitely not required, its just a bit more work on the part of the release engineer who has to know to manually run the new regions.  Its a bit more error prone and less guaranteed.  

As for the API call I'm not sure how it would work exactly?  What do we want the scrapers writers to do when wanting a region to be released?  If they just update the cron and nothing else I think it'll be pretty hard to programmatically compare that to the older version and see which regions were added?  I don't know off the top of my head if you can easily say 'compare the contents of this file from last weeks version'

jessex commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
+over the last two days.  Check the monitoring page to see if anything is fishy.
+2.  If all looks good, be sure to move the successful scrapers from [staging cron](staging/cron.yaml) to
+the [prod cron](cron.yaml) file.  Don't move the test cron as we do not want that running prod.  This is an important step, we do not want to run the jobs in two places!
+3.  Push that change to the `staging` branch
```
Missing a period here.

arian487 replied:
Fixed.

jessex commented on README.md:
```diff
@@ -116,19 +116,38 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+#### On Monday Morning
+
+1. Verify that the tests are all passing in [Travis](https://travis-ci.org/Recidiviz/recidiviz-data)
```
Missing a period here.

arian487 replied:
Done

colincadams commented:
I haven't read through the whole conversation, but can we use tags instead of branches create versions for release. Then we don't have to deal with the merging back and forth?

arian487 commented:
I don't think travis supports deploys from tags, so we have to use branches.

arian487 commented:
I've updated the process and documentation as per our discussion @jessex.  I threw together an extremely simply deploy script that the release engineer can run but I'm happy to take suggestions on how that should be structured.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+#### On Monday Morning
+
+1. Verify that the tests are all passing in [Travis](https://travis-ci.org/Recidiviz/recidiviz-data).
+2. The release engineer should tag a commit with "va.b.c".  For now just increment the numbers until a better method is in place [TODO #646](https://github.com/Recidiviz/recidiviz-data/issues/646) This will trigger a release to staging.
```
Need a period after the TODO.

arian487 replied:
Done.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+2. The release engineer should tag a commit with "va.b.c".  For now just increment the numbers until a better method is in place [TODO #646](https://github.com/Recidiviz/recidiviz-data/issues/646) This will trigger a release to staging.
+3. Once the release is complete, run `https://recidiviz-staging.appspot.com/scraper/start?region=us_fl_martin` [TODO #623](https://github.com/Recidiviz/recidiviz-data/issues/623)
+and verify that it is happy by looking at the monitoring page [TODO #59](https://github.com/Recidiviz/recidiviz-data/issues/59) and also checking
+the logs for errors.
```
I think you can push this back to the end of the previous line.

arian487 replied:
Done.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+4. Manually check to see which set of scrapers have been added to the cron.yaml since last release and manually run all of those scrapers as well.
+5. Every morning until Wednesday, rerun all of the to be deployed scrapers manually and make sure there aren't any failures.
+6. For the next two days periodically check to make sure the build is happy and monitoring is good for all of the regions.  If there are errors try to fix them or
+contact the scraper writer to fix them.
```
This can go back to the end of the previous line, too.

arian487 replied:
Done.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+#### On Wednesday Morning
+
+1.  Verify that the scraper crons as well as the `infer_release` cron ran successfully
```
"...ran successfully _in staging_ over the last two days..."

arian487 replied:
Done.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+1.  Verify that the scraper crons as well as the `infer_release` cron ran successfully
+over the last two days.  Check the monitoring page to see if anything is fishy.
+3.  Once you are ready to release to production, run `./deploy_production <release_tag>` to finally release the code to production.
```
Maybe append this number with, "This will checkout recent tags and deploy the most recent one automatically."

arian487 replied:
I updated the comment, but the script as is does not deploy the most recent one, I figured the user should have to explicitly provide a release tag.

jessex commented on README.md:
```diff
@@ -116,19 +116,33 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+#### On Monday Morning
+
+1. Verify that the tests are all passing in [Travis](https://travis-ci.org/Recidiviz/recidiviz-data).
```
That will help because the second list looks mis-numbered as is.

arian487 replied:
Oh, thats an excellent protip, thanks!  Done.

jessex commented on deploy_production.sh:
```diff
@@ -0,0 +1,16 @@
+echo "Starting deploy of cron.yaml"
+gcloud app deploy cron.yaml --project=recidiviz-123
+
+echo "Starting deploy of queues.yaml"
```
queue.yaml, singular

arian487 replied:
Done.

arian487 replied:
Filed an issue: https://github.com/Recidiviz/recidiviz-data/issues/649

I'll say that I think its risky to have the same script handle both staging and production deploy.  If we guard it with a flag for example, the user might have intended to do staging but accidentally do production.  Personally I think its safer to have a strong separation by just having them be different scripts so the user has to very deliberately run a deploy production script vs a deploy staging script.

jessex commented on deploy_production.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
I know that we're going with auto-deploys to stage and manual deploys to prod for now, but I could see it being useful to toggle this script to manually deploy to stage, as well. If not necessary yet, we could file that away as a future enhancement.

arian487 replied:
Filed an issue: #649

I'll say that I think its risky to have the same script handle both staging and production deploy. If we guard it with a flag for example, the user might have intended to do staging but accidentally do production. Personally I think its safer to have a strong separation by just having them be different scripts so the user has to very deliberately run a deploy production script vs a deploy staging script.

jessex commented on recidiviz/staging/cron.yaml:
```diff
@@ -0,0 +1,33 @@
+    job_age_limit: 9h
+
+
+# Scrapers to be promoted to prod
```
Can remove this line safely now.

arian487 replied:
Done.

arian487 commented:
I've addressed the comments, thanks for taking a look.  Once its in, I'll run through the release process myself (I'll do it over two hours instead of two days), to make sure it makes sense and works.  We'll have it in place for next weeks release.

colincadams commented on recidiviz/staging/cron.yaml:
```diff
@@ -0,0 +1,30 @@
+# LEAVE THESE HERE, THESE ARE TEST SCRAPERS AND JOBS FOR STAGING.  We are
```
This file is no longer necessary correct?

arian487 replied:
I think we should keep it since the cron for staging and prod is different, and we should have a file for it.

colincadams replied:
Okay then should it be pushed automatically by Travis?

colincadams commented on .travis.yml:
```diff
@@ -1,26 +1,44 @@
 - mypy recidiviz
 after_success:
 - coveralls
+deploy:
```
@zdg2102 Is there a way to make this run alembic migrations as well?

colincadams commented on deploy_production.sh:
```diff
@@ -0,0 +1,16 @@
+
+echo "Checking out tag $1"
+git checkout tags/$1 -b $1
+
```
@zdg2102 @arian487 and run alembic here as well?

arian487 replied:
We can think about how alembic fits into this a bit later I hadn't thought about that.

colincadams replied:
ack, filed #657

arian487 commented:
My thoughts on more frequent releases is that I think the end state will definitely not have more releases, and in this phase of very active development, I'm happy running releases piecemeal like we have been now (following this process still).  I don't want to add too much process while we have a lot of work to do.

arian487 commented:
Gentle nudge on this PR, I'd like to get it in to try the full process before drop day tomorrow

colincadams commented on README.md:
```diff
@@ -116,19 +116,31 @@ described in [running tests](#running-tests)). In practice, this is not particul
+
+The release engineer oncall should go through the following steps:
+
+#### On Monday Morning
```
re: frequent releases. Agreed we should use this process for more frequent releases, so I would just rename these headers to *Push to Staging* and *Push to Prod* so that the docs are still relevant. Then in the text you can say 'typically on Monday morning' if preferred.

arian487 replied:
Done.
